src/google/protobuf/generated_message_tctable_impl.h

```diff
@@ -795,17 +803,27 @@ class PROTOBUF_EXPORT TcParser final {
                       &msg->_internal_metadata_, ctx);
     } else {
       // Otherwise, we directly put it on the unknown field set.
-      return UnknownFieldParse(
-          tag,
-          msg->_internal_metadata_.mutable_unknown_fields<UnknownFieldsT>(),
-          ptr, ctx);
+#ifdef BYTEDANCE_PROTOBUF_SKIP_PARSING_UNKNOWN_FIELD
+      bool skip_unknown = true;
+#else
+      bool skip_unknown = false;
+#endif
+      if (skip_unknown) {
+        // nullptr is special cased to indicate skipping
+        return UnknownFieldParse(tag, static_cast<std::string*>(nullptr), ptr, ctx);
+      } else {
+        return UnknownFieldParse(
+            tag,
+            msg->_internal_metadata_.mutable_unknown_fields<UnknownFieldsT>(),
+            ptr, ctx);
+      }
     }
   }
 
   template <class MessageBaseT>
   static const char* MessageSetWireFormatParseLoopImpl(
       PROTOBUF_TC_PARAM_NO_DATA_DECL) {
-    return RefAt<ExtensionSet>(msg, table->extension_offset)
+    return RefAt<ExtensionSet>(msg, table->extension_offset & -2)
```

Comment:
Why do we need this bit masking? Would it work with and without new parser?

Reply:
In order to support message set natively (which was introduced in main line recently). This was introduced by a special 1 entry fast path handler that would branch to MessageSetWireFormatParseLoopImpl. The new table doesn't store function pointers so I added message set in the fallback parse function. I use the lowest bit of the offset of the extension set (which due to alignment is always 0), to indicate if this proto is a message set. In which case the parser just calls the fallback, which will call this function.